components/upload-progress-fullpage.tsx

```diff
@@ -74,11 +77,53 @@ const UploadProgressFullpage: React.FC<Props> = ({ file }) => {
     }
   }, [upload]);
 
+
+  const startFileValidation = (_file: File) => {
+    // Attempt to load the file as a video element and inspect its duration
+    // metadata. This is not an authoritative check of video duration, but
+    // rather intended to serve as just a simple and fast sanity check.
+    if (!_file.type.includes("video")) {
+      console.warn(`file type (${_file.type}) does not look like video!`);
+      setInputValidated(true);
+      return;
+    }
+
+    const video = document.createElement('video');
+    video.preload = 'metadata';
+    video.onloadedmetadata = function() {
+      URL.revokeObjectURL(video.src);
+      if (video.duration > MAX_VIDEO_DURATION_SEC) {
+        setValidationError(`file duration (${video.duration.toString()}s) exceeds allowed maximum (${MAX_VIDEO_DURATION_SEC}s)!`);
+        return;
+      }
+      setInputValidated(true);
+    };
+    video.onerror = function() {
+      // The file has a video MIME type, but we were unable to load its
+      // metadata for some reason.
+      console.warn("failed to load video file metadata for validation!");
+      URL.revokeObjectURL(video.src);
+      setInputValidated(true);
+    };
+    video.src = URL.createObjectURL(file);
+  };
+
   useEffect(() => {
-    if (file) {
+    if (!file) {
+      return;
+    }
+
+    if (inputValidated || validationError) {
+      if (validationError) {
+        setErrorMessage(validationError);
+        return;
+      }
       startUpload(file);
+      return;
     }
-  }, [file]);
+
+    startFileValidation(file);
```

Comment:
I think a better pattern here would be to make `startFileValidation` return a promise.

- no longer need `inputValidated` boolean state
- no longer need `validationError` state

it would be more like this:

```js
useEffect(() => {
  if (!file) {
    return
  }
  startFileValidation(file).then(() => {
   startUpload(file);
}).catch((error) => {
  setErrorMessage(error.message);
})
}, [file])
```

I think that would clean things up a bit and make the code a little more readable/easy to follow

Reply:
Nice, thanks! This looks like a much nicer way to do this